### PR TITLE
[ADAM-1457] Add support for loading BGZF compressed FASTQ in parallel.

### LINKED ISSUE #1457: Splittable compression and FASTQ
[This](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java#L131) exception should check to make sure that the compression codec is not splittable, and then we need to make sure that the rest of the code works on splittable compression codecs.

### PR DESCRIPTION
**Pull request by fnothaft**
WIP. Resolves #1457. Pending added unit tests and further validation and in-line docs.

### CODE REVIEW

File: adam-core/src/main/java/org/bdgenomics/adam/io/FastqInputFormat.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.seqdoop.hadoop_bam.util.BGZFEnhancedGzipCodec;
+
+public abstract class FastqInputFormat extends FileInputFormat<Void, Text> {
+
+    protected boolean splittable;
+
+    /**
+     * For now, we do not support splittable compression codecs. As in, we will
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/FastqInputFormat.java]: Update doc when this pull request fixes #1457

File: adam-core/src/main/java/org/bdgenomics/adam/io/FastqInputFormat.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     * read the compressed data, but we will not allow it to be splittable. We
+     * will fix this in #1457.
+     *
+     * @param context The job context to get the configuration from.
+     * @param filename The path the input file is saved at.
+     * @return Returns false if this file is compressed.
+     */
+    @Override protected boolean isSplitable(JobContext context, Path filename) {
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/FastqInputFormat.java]: final method parameters

File: adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java [hunk 3]
```diff
@@ -105,13 +107,31 @@
     private static final int MAX_LINE_LENGTH = 10000;
 
     /**
+     * True if the underlying data is compressed with a splittable codec.
+     */
+    private boolean compressedAndSplittable = false;
+    
+    /**
+     * True if the last read was <=0 bytes in size.
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java]: `<` breaks javadoc, use `&lt;` or better yet, isn't size always at least zero?

File: adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java [hunk 4]
```diff
@@ -152,15 +182,22 @@ protected FastqRecordReader(final Configuration conf, final FileSplit split) thr
...
      * @param stream The stream to reposition.
      */
-    protected final void positionAtFirstRecord(final FSDataInputStream stream) throws IOException {
+    protected final int positionAtFirstRecord(final FSDataInputStream stream,
+                                              final CompressionCodec codec) throws IOException {
         Text buffer = new Text();
-
+        long originalStart = start;
+        
         if (true) { // (start > 0) // use start>0 to assume that files start with valid data
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java]: while yer in here, could you take out this loop?

File: adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.compress.SplitCompressionInputStream;
+import org.apache.hadoop.mapreduce.lib.input.CompressedSplitLineReader;
+
+class ResettableCompressedSplitLineReader extends CompressedSplitLineReader {
+
+    private boolean isFinished = false;
+    private boolean needsAdditionalRecord = false;
+    private SplitCompressionInputStream sin;
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java]: final field

File: adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    
+    public ResettableCompressedSplitLineReader(SplitCompressionInputStream in,
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java]: final method params

File: adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        super(in, conf, null);
+        sin = in;
+    }
+    
+    public void reset() {
+        isFinished = false;
+    }
+    
+    @Override
+    public int readLine(Text str, int maxLineLength, int maxBytesToConsume)
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java]: final method params

File: adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        return bytesRead;
+    }
+    
+    @Override
+    public boolean needAdditionalRecordAfterSplit() {
+        return !isFinished && needsAdditionalRecord;
+    }
+
+    @Override
+    protected int fillBuffer(InputStream in, byte[] buffer, boolean inDelimiter)
```
Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/ResettableCompressedSplitLineReader.java]: final method params

File: adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala [hunk 9]
```diff
@@ -63,7 +63,11 @@ private[adam] object FileExtensions {
    */
   def isFastqExt(pathName: String): Boolean = {
     pathName.endsWith(".fq") ||
-      pathName.endsWith(".fastq")
+      pathName.endsWith(".fastq") ||
+      pathName.endsWith(".fastq.gz") ||
+      pathName.endsWith(".fastq.bgz")
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala]: typo here, missing `||`

File: adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala [hunk 10]
```diff
@@ -63,7 +63,11 @@ private[adam] object FileExtensions {
    */
   def isFastqExt(pathName: String): Boolean = {
     pathName.endsWith(".fq") ||
-      pathName.endsWith(".fastq")
+      pathName.endsWith(".fastq") ||
+      pathName.endsWith(".fastq.gz") ||
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala]: We removed compressed file extensions from `FileExtensions` in review for commit 93d17b418ddc918ae95095b961ae7ca045be3f5d.  There is an [`ADAMContext.trimExtensionIfCompressed`](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala#L1281) method that checks the Hadoop configuration.  Can that still work with these changes?

File: adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java (hunk 3 above)
Comment (fnothaft) [inline adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java]: Alas, the splittable codecs use -2 to signify end of splittable block and -1 to signify end of stream. I'll fix the `<` though.

Comment (heuermh) [inline adam-core/src/main/java/org/bdgenomics/adam/io/FastqRecordReader.java]: Awesome, that makes my layover.

**Comment by fnothaft** [comment]
Added unit tests and addressed review comments. This is good to go from my side.

**Comment by fnothaft** [comment]
@heuermh you can only imagine how much fun I've had...

File: adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala (hunk 10 above)
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala]: On a related note, the doc [here](https://github.com/fnothaft/adam/blob/cefab0105d1948c9724ed0eb88617dc6e8de0657/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala#L2646) and similar in related methods should be updated.  Is bz2 still supported?  Other user-configured Hadoop compression codecs?

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala]: Yeah, good catch RE docs update and `ADAMContext.trimExtensionIfCompressed`. Due to the vagaries of how the BGZF codecs get set up, I'll need to modify `ADAMContext.trimExtensionIfCompressed` but it is a very small modification.

I'll also make a more extensive documentation pass.

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/util/FileExtensions.scala]: Actually, didn't even need a modification to `ADAMContext.trimExtensionIfCompressed`.

**Comment by fnothaft** [comment]
Addressed the remaining review feedback and added some inline docs in the hairiest parts of this PR.

**Comment by devin-petersohn** [comment]
Thanks @fnothaft!

**Pull request merged**